### PR TITLE
Support for reading dkg summary from a previous run as genesis dkg.

### PR DESCRIPTION
**Pull request by Kenwes13**
## Premise
Genesis miners need to have the appropriate dkg secret shares that correspond to the magic block's mpks in the genesis block. To make the process easier the miners will join an already running chain. Once all the miners join and go through one view change together the magic block from that view change can be downloaded and used as the magic block for the genesis. The miners and sharders will stop and be sent the genesis magic block which allows the corresponding saved dkg summary to be used as the genesis dkg. The miners just need to use the original magic block number for the genesis_dkg in the 0chain.yaml.

## Changes
- Added a verify function for the dkg summary to be used the the mpks.
- Updated miner to read the dkg summary from a json when the genesis_dkg is zero. 
- Updated miner to read the dkg summary from the database if the genesis_dkg isn't zero.
- Updated non genesis miners and sharders to read their description from their keys file.
- Updated the docker.local/bin/clean.sh to exclude the miners rocksdb for their dkgs.
- Added docker.local/bin/clean_miners_dkg.sh to clean the rocksdb when needed.
- Updated the miner b0docker-compose to include the miners dkg.json file instead of reading it from the magic block.
- Added genesis_dkg under the network in the 0chain.yaml.
- Replaced the magic block in b0magicBlock_4_miners_1_sharder.json with a new one that doesn't have the secret shares for all the miners.
- Added the dkg summary jsons for miners 1-4.
- Deleted the msk files for miners 1-4.
- Updated miners 5-8 key files to include a description.
- Updated sharders 2 and 3 key files to include a description.

### CODE REVIEW

**Comment by Sriep** [comment]
I added support to `0chain/local` for the change. Added a 1 sharder 3 miner magic block for developers with only 3 PCs, and updated readmes.

File: code/go/0chain.net/chaincore/threshold/bls/dkg.go [hunk 1]
```diff
@@ -357,14 +357,17 @@ func SetupDKGSummary(store datastore.Store) {
 
 func SetupDKGDB() {
 	//Clean old files on restart
```
Comment (Sriep) [inline code/go/0chain.net/chaincore/threshold/bls/dkg.go:359]: I think the orphaned comment can go?

**Review by Sriep — COMMENTED** [review]
Could you explain ` The miners just need to use the original magic block number for the genesis_dkg in the 0chain.yaml.` in more detail. Where do I find the `original magic block number`?

**Comment by Sriep** [comment]
Should any of the `readme.md` files be updated with information about this change?

**Comment by Kenwes13** [comment]
Oh awesome! 

Yea, the plan is to have all the outside miners join a running network with the view change enabled. Once all the miners join we will save that magic block and change some of the values like the magic block number will be set to 1 and round number will be set to 0. It will be rehashed and then use that as the genesis magic block for a new network. The mpks in that magic block need to match the dkg summaries that were saved from the last run. So before the change of the magic block the magic block number, which is used as the ID for the dkg summary, needs to be noted. That value will be used as the genesis_dkg so the dkg summary matches genesis the genesis mpks.

Yes. I'll update the readme.md that talks about the creation of the magic block for the genesis. I'll also update the code to output the secret shares used by the miners and include that as well. I think that should cover it. What do you think?

**Review by peterlimg — CHANGES_REQUESTED** [review]
I tested to run BC from scratch locally for this, but seems it can not work, the BC did not move and I saw repeated errors of 
```
ERROR	miner/protocol_bls.go:249	failed to verify share	{"share": "01120f30675e84f3aed68dc8a17c4970379a813b835477a45bcd1e967db02f89", "message": "127ba73263f8a16b41", "from": "131810bd1258ae95955fb40c7ef72498", "pi": "b8ca62305fa283ebe2e86137e0e82f0f84f41b162ae6e51e10032c6a5a66332339824ae8c7c2fa3a1e8a23521090c9997dcb6f38f22738a55068bb608f539e04", "node_id": "31810bd1258ae95955fb40c7ef72498a556d3587121376d9059119d280f34929", "round": 1, "dkg_starting_round": 0, "dkg_mb_number": 1, "chain_tc": 28, "round_tc": 27, "pr_seed": "ba73263f8a16b41"}
```
and 
```
ERROR	chain/protocol_view_change.go:597	get_dkg_phase_from_sharders -- no phases given
```

File: code/go/0chain.net/miner/miner/main.go [hunk 2]
```diff
@@ -224,15 +226,33 @@ func main() {
...
-		for k, v := range mb.GetShareOrSigns().GetShares() {
-			dkgShare.SecretShares[miner.ComputeBlsID(k)] = v.ShareOrSigns[node.Self.Underlying().GetKey()].Share
+		if genesisDKG == 0 {
+			oldDKGShare, err = miner.ReadDKGSummaryFile(*dkgFile)
+			if err != nil {
+				Logger.Panic(fmt.Sprintf("Error reading DKG file. ERROR: %v", err.Error()))
+			}
+		} else {
+			oldDKGShare, err = miner.LoadDKGSummary(ctx, strconv.FormatInt(genesisDKG, 10))
+			if err != nil && !config.DevConfiguration.ViewChange {
```
[the remark this answers is not in the document]
Comment (Kenwes13) [reply code/go/0chain.net/miner/miner/main.go:241]: Can you tell me the steps to reproduce the error?

**Comment by Kenwes13** [comment]
The create magic block has been updated to create the dkg summaries for the miners so they match the magic block mpks. The readme has also been updated.

File: code/go/0chain.net/miner/miner/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (peterlimg) [reply code/go/0chain.net/miner/miner/main.go:241]: I mean the `err` variable is not logged if `config.DevConfiguration.ViewChange` is true.

[the remark this answers is not in the document]
Comment (Kenwes13) [reply code/go/0chain.net/miner/miner/main.go:241]: Oh! Yes, that needs to be changed! What about the block chain not moving issue?

[the remark this answers is not in the document]
Comment (peterlimg) [reply code/go/0chain.net/miner/miner/main.go]: Great. 

Hmm, still not work for both VC:false and VC:true. I'm using the default 4_miners_1_sharder.json file. All settings in 0chain.yaml and sc.yaml are default, except when view_change is true. Do you have changes on the magicblock json or config yaml files that are not commited. Or maybe you can have a test with default settings.

[the remark this answers is not in the document]
Comment (guruhubb) [reply code/go/0chain.net/miner/miner/main.go]: Also please test with https://github.com/0chain/0chain/tree/master/quickstart/1s_2m_4b

[the remark this answers is not in the document]
Comment (Kenwes13) [reply code/go/0chain.net/miner/miner/main.go]: No, all the changes have been committed. I've just tested it with VC as both true and false, and the blockchain seems to be running smoothly for both. What are the errors you're getting?

[the remark this answers is not in the document]
Comment (anish-squareops) [reply code/go/0chain.net/miner/miner/main.go]: @guruhubb @Kenwes13 @peterlimg I gave it a try and it worked fine with vc: false. With vc: true, sharder2 joined the chain immediately but miner5 joined after 1000-1500 blocks so i tried joining miner6 and the chain got stuck. I also 
 noticed external miner / sharder urls are https://localhost/miner75/_diagnostics / https://localhost/sharder72/_diagnostics, which are not correct for local setup.

[the remark this answers is not in the document]
Comment (Kenwes13) [reply code/go/0chain.net/miner/miner/main.go]: I think the block chain getting stuck with miner 5 and 6 might have to do with a deadlock in the SetNodeInfo. There is a PR to fix that issue. The paths need to be delete in the b0mnode*N*_keys.txt where *N* is the number of the miner folder to use for local.

**Comment by peterlimg** [comment]
Works for me after deleting the previously generated miner Dockerfile.

**Pull request merged**